processing/src/main/java/org/apache/druid/segment/join/HashJoinSegment.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import org.apache.druid.segment.AbstractSegment;
+import org.apache.druid.segment.QueryableIndex;
+import org.apache.druid.segment.Segment;
+import org.apache.druid.segment.StorageAdapter;
+import org.apache.druid.timeline.SegmentId;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Represents a deep, left-heavy join of a left-hand side baseSegment onto a series of right-hand side clauses.
+ *
+ * In other words, logically the operation is: join(join(join(baseSegment, clauses[0]), clauses[1]), clauses[2]) etc.
+ */
+public class HashJoinSegment extends AbstractSegment
+{
+  private final Segment baseSegment;
+  private final List<JoinableClause> clauses;
+
+  public HashJoinSegment(
+      Segment baseSegment,
+      List<JoinableClause> clauses
+  )
+  {
+    this.baseSegment = baseSegment;
+    this.clauses = clauses;
+  }
+
+  @Override
+  public SegmentId getId()
+  {
+    return baseSegment.getId();
+  }
+
+  @Override
+  public Interval getDataInterval()
+  {
+    // __time column will come from the baseSegment, so use its data interval.
+    return baseSegment.getDataInterval();
+  }
+
+  @Nullable
+  @Override
+  public QueryableIndex asQueryableIndex()
+  {
+    // Even if baseSegment is a QueryableIndex, we don't want to expose it, since we've modified its behavior
+    // too much while wrapping it.
+    return null;
+  }
```

Comment:
I would add a unit test since this is an override, with a specific need to ensure correctness in the future.